File: libs/sdk-java/src/main/java/io/daytona/sdk/SandboxProcess.java
// Copyright Daytona Platforms Inc.
// SPDX-License-Identifier: Apache-2.0

package io.daytona.sdk;

import io.daytona.sdk.model.*;
import io.daytona.sdk.exception.DaytonaException;
import io.daytona.toolbox.client.api.ProcessApi;
import io.daytona.toolbox.client.model.CreateSessionRequest;
import io.daytona.toolbox.client.model.ExecuteRequest;
import io.daytona.toolbox.client.model.PtyCreateRequest;
import io.daytona.toolbox.client.model.PtyCreateResponse;
import io.daytona.toolbox.client.model.PtyResizeRequest;
import okhttp3.Request;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class SandboxProcess {
    private final ProcessApi processApi;
    private final Sandbox sandbox;

    SandboxProcess(ProcessApi processApi, Sandbox sandbox) {
        this.processApi = processApi;
        this.sandbox = sandbox;
    }

    public ExecuteResponse executeCommand(String command) {
        return executeCommand(command, null, null, null);
    }

    public ExecuteResponse executeCommand(String command, String cwd, Map<String, String> env, Integer timeout) {
        ExecuteRequest request = new ExecuteRequest().command(command);
        if (cwd != null) {
            request.cwd(cwd);
        }
        if (timeout != null) {
            request.timeout(timeout);
        }
        io.daytona.toolbox.client.model.ExecuteResponse response = ExceptionMapper.callToolbox(() -> processApi.executeCommand(request));
        return toExecuteResponse(response);
    }

    public ExecuteResponse codeRun(String code) {
        String lang = "python";
        if (sandbox.getLabels() != null && sandbox.getLabels().containsKey("code-toolbox-language")) {
            lang = sandbox.getLabels().get("code-toolbox-language");
        }

        String scriptPath;
        String runCmd;
        if ("typescript".equalsIgnoreCase(lang)) {
            scriptPath = "/tmp/daytona_code.ts";
            runCmd = "npx ts-node " + scriptPath;
        } else if ("javascript".equalsIgnoreCase(lang)) {
            scriptPath = "/tmp/daytona_code.js";
            runCmd = "node " + scriptPath;
        } else {
            scriptPath = "/tmp/daytona_code.py";
            runCmd = "python3 " + scriptPath;
        }

        String command = "cat <<'DAYTONA_EOF' > " + scriptPath + "\n"
                + code + "\nDAYTONA_EOF\n"
                + runCmd;

        return executeCommand(command);
    }

    public void createSession(String sessionId) {
        ExceptionMapper.runToolbox(() -> processApi.createSession(new CreateSessionRequest().sessionId(sessionId)));
    }

    public Session getSession(String sessionId) {
        io.daytona.toolbox.client.model.Session session = ExceptionMapper.callToolbox(() -> processApi.getSession(sessionId));
        return toSession(session);
    }

    public SessionExecuteResponse executeSessionCommand(String sessionId, SessionExecuteRequest req) {
        io.daytona.toolbox.client.model.SessionExecuteRequest request = new io.daytona.toolbox.client.model.SessionExecuteRequest()
                .command(req.getCommand())
                .runAsync(req.getRunAsync());
        io.daytona.toolbox.client.model.SessionExecuteResponse response = ExceptionMapper.callToolbox(() -> processApi.sessionExecuteCommand(sessionId, request));
        return toSessionExecuteResponse(response);
    }

    public Command getSessionCommand(String sessionId, String commandId) {
        io.daytona.toolbox.client.model.Command command = ExceptionMapper.callToolbox(() -> processApi.getSessionCommand(sessionId, commandId));
        return toCommand(command);
    }

    public String getSessionCommandLogs(String sessionId, String commandId) {
        return ExceptionMapper.callToolbox(() -> processApi.getSessionCommandLogs(sessionId, commandId, null));
    }

    public void deleteSession(String sessionId) {
        ExceptionMapper.runToolbox(() -> processApi.deleteSession(sessionId));
    }

    public List<Session> listSessions() {
        List<io.daytona.toolbox.client.model.Session> sessions = ExceptionMapper.callToolbox(processApi::listSessions);
        List<Session> output = new ArrayList<Session>();
        if (sessions != null) {
            for (io.daytona.toolbox.client.model.Session session : sessions) {
                output.add(toSession(session));
            }
        }
        return output;
    }

    public PtyHandle createPty(PtyCreateOptions options) {
        PtyCreateOptions createOptions = options == null ? new PtyCreateOptions() : options;
        PtyCreateRequest request = new PtyCreateRequest()
                .id(createOptions.getId())
                .cols(createOptions.getCols())
                .rows(createOptions.getRows());

        PtyCreateResponse response = ExceptionMapper.callToolbox(() -> processApi.createPtySession(request));
        if (response == null || response.getSessionId() == null || response.getSessionId().isEmpty()) {
            throw new DaytonaException("Failed to create PTY session");
        }

        String sessionId = response.getSessionId();
        String wsUrl = buildPtyWebSocketUrl(sandbox.getToolboxApiClient().getBasePath(), sessionId);
        Request wsRequest = new Request.Builder()
                .url(wsUrl)
                .addHeader("Authorization", "Bearer " + sandbox.getApiKey())
                .build();

        return new PtyHandle(
                sandbox.getToolboxApiClient().getHttpClient(),
                wsRequest,
                sessionId,
                this::resizePtySession,
                this::killPtySession,
                createOptions.getOnData()
        );
    }

    public void resizePtySession(String sessionId, int cols, int rows) {
        ExceptionMapper.callToolbox(() -> processApi.resizePtySession(
                sessionId,
                new PtyResizeRequest().cols(cols).rows(rows)
        ));
    }

    public void killPtySession(String sessionId) {
        ExceptionMapper.callToolbox(() -> processApi.deletePtySession(sessionId));
    }

    private String buildPtyWebSocketUrl(String toolboxBaseUrl, String sessionId) {
        if (toolboxBaseUrl == null || toolboxBaseUrl.isEmpty()) {
            throw new DaytonaException("Toolbox base URL is not available");
        }
        String wsBase = toolboxBaseUrl
                .replaceFirst("^https://", "wss://")
                .replaceFirst("^http://", "ws://");
        return wsBase + "/process/pty/" + sessionId + "/connect";
    }

    private ExecuteResponse toExecuteResponse(io.daytona.toolbox.client.model.ExecuteResponse source) {
        ExecuteResponse response = new ExecuteResponse();
        if (source != null) {
            response.setExitCode(source.getExitCode());
            response.setResult(source.getResult());
        }
        return response;
    }

    private Session toSession(io.daytona.toolbox.client.model.Session source) {
        Session session = new Session();
        if (source == null) {
            return session;
        }
        session.setSessionId(source.getSessionId());
        List<Command> commands = new ArrayList<Command>();
        if (source.getCommands() != null) {
            for (io.daytona.toolbox.client.model.Command command : source.getCommands()) {
                commands.add(toCommand(command));
            }
        }
        session.setCommands(commands);
        return session;
    }

    private Command toCommand(io.daytona.toolbox.client.model.Command source) {
        Command command = new Command();
        if (source != null) {
            command.setId(source.getId());
            command.setCommand(source.getCommand());
            command.setExitCode(source.getExitCode());
        }
        return command;
    }

    private SessionExecuteResponse toSessionExecuteResponse(io.daytona.toolbox.client.model.SessionExecuteResponse source) {
        SessionExecuteResponse response = new SessionExecuteResponse();
        if (source != null) {
            response.setCmdId(source.getCmdId());
            response.setOutput(source.getOutput());
            response.setExitCode(source.getExitCode());
        }
        return response;
    }
}

File: libs/sdk-dotnet/SandboxProcess.cs
// Copyright Daytona Platforms Inc.
// SPDX-License-Identifier: Apache-2.0

using System.Net.WebSockets;
using System.Text;
using Daytona.Sdk.Models;
using CreateSessionRequestModel = Daytona.ToolboxApiClient.Model.CreateSessionRequest;
using ExecuteRequestModel = Daytona.ToolboxApiClient.Model.ExecuteRequest;
using ProcessApi = Daytona.ToolboxApiClient.Api.ProcessApi;
using PtyCreateRequestModel = Daytona.ToolboxApiClient.Model.PtyCreateRequest;
using PtyResizeRequestModel = Daytona.ToolboxApiClient.Model.PtyResizeRequest;
using SessionExecuteRequestModel = Daytona.ToolboxApiClient.Model.SessionExecuteRequest;

namespace Daytona.Sdk;

public class SandboxProcess
{
    private static readonly byte[] StdoutPrefix = [0x01, 0x01, 0x01];
    private static readonly byte[] StderrPrefix = [0x02, 0x02, 0x02];

    private readonly Sandbox _sandbox;
    private readonly ProcessApi _api;

    internal SandboxProcess(Sandbox sandbox, ProcessApi api)
    {
        _sandbox = sandbox;
        _api = api;
    }

    public async Task<ExecuteResponse> ExecuteCommandAsync(
        string command,
        string? cwd = null,
        Dictionary<string, string>? env = null,
        int? timeout = null,
        CancellationToken ct = default)
    {
        var effectiveCommand = MergeEnvironmentWithCommand(command, env);
        var response = await GeneratedClientSupport.ExecuteToolboxAsync(
            () => _api.ExecuteCommandAsync(new ExecuteRequestModel(effectiveCommand, cwd, timeout ?? 10), ct),
            ct
        );

        return GeneratedClientSupport.ToSdkExecuteResponse(response);
    }

    public async Task<ExecuteResponse> CodeRunAsync(string code, CancellationToken ct = default)
    {
        var language = _sandbox.Labels.TryGetValue("code-toolbox-language", out var value)
            ? value
            : "python";

        var extension = language switch
        {
            "javascript" => "js",
            "typescript" => "ts",
            _ => "py"
        };

        var interpreter = language switch
        {
            "javascript" => "node",
            "typescript" => "npx ts-node",
            _ => "python3"
        };

        var token = $"DAYTONA_EOF_{Guid.NewGuid():N}";
        var tempFile = $"/tmp/daytona_code.{extension}";
        var command = $"cat <<'{token}' > {tempFile}\n{code}\n{token}\n{interpreter} {tempFile}";
        return await ExecuteCommandAsync(command, timeout: 60, ct: ct);
    }

    public Task CreateSessionAsync(string sessionId, CancellationToken ct = default)
        => GeneratedClientSupport.ExecuteToolboxAsync(
            () => _api.CreateSessionAsync(new CreateSessionRequestModel(sessionId), ct),
            ct
        );

    public async Task<Session> GetSessionAsync(string sessionId, CancellationToken ct = default)
    {
        var session = await GeneratedClientSupport.ExecuteToolboxAsync(
            () => _api.GetSessionAsync(sessionId, ct),
            ct
        );
        return GeneratedClientSupport.ToSdkSession(session);
    }

    public async Task<SessionExecuteResponse> ExecuteSessionCommandAsync(
        string sessionId,
        SessionExecuteRequest req,
        CancellationToken ct = default)
    {
        var response = await GeneratedClientSupport.ExecuteToolboxAsync(
            () => _api.SessionExecuteCommandAsync(
                sessionId,
                new SessionExecuteRequestModel(
                    async: false,
                    command: req.Command,
                    runAsync: req.RunAsync ?? false,
                    suppressInputEcho: req.SuppressInputEcho ?? false
                ),
                ct
            ),
            ct
        );

        var mapped = GeneratedClientSupport.ToSdkSessionExecuteResponse(response);

        if (string.IsNullOrEmpty(mapped.Output))
        {
            return mapped;
        }

        var demuxed = DemuxOutput(mapped.Output);
        return mapped with
        {
            Stdout = string.IsNullOrEmpty(mapped.Stdout) ? demuxed.Stdout : mapped.Stdout,
            Stderr = string.IsNullOrEmpty(mapped.Stderr) ? demuxed.Stderr : mapped.Stderr
        };
    }

    public async Task<Command> GetSessionCommandAsync(string sessionId, string commandId, CancellationToken ct = default)
    {
        var command = await GeneratedClientSupport.ExecuteToolboxAsync(
            () => _api.GetSessionCommandAsync(sessionId, commandId, ct),
            ct
        );
        return GeneratedClientSupport.ToSdkCommand(command);
    }

    public async Task<string> GetSessionCommandLogsAsync(string sessionId, string commandId, CancellationToken ct = default)
    {
        var output = await GeneratedClientSupport.ExecuteToolboxAsync(
            () => _api.GetSessionCommandLogsAsync(sessionId, commandId, cancellationToken: ct),
            ct
        );

        var demux = DemuxOutput(output);
        return string.IsNullOrEmpty(demux.Stdout) && string.IsNullOrEmpty(demux.Stderr)
            ? output
            : demux.Stdout + demux.Stderr;
    }

    public Task DeleteSessionAsync(string sessionId, CancellationToken ct = default)
        => GeneratedClientSupport.ExecuteToolboxAsync(
            () => _api.DeleteSessionAsync(sessionId, ct),
            ct
        );

    public async Task<PtyHandle> CreatePtyAsync(PtyCreateOptions options, CancellationToken ct = default)
    {
        await GeneratedClientSupport.ExecuteToolboxAsync(
            () => _api.CreatePtySessionAsync(new PtyCreateRequestModel(
                cols: options.Cols,
                rows: options.Rows,
                id: options.Id
            ), ct),
            ct
        );

        var baseUrl = _sandbox.ToolboxBaseUrl;
        var wsUrl = baseUrl
            .Replace("https://", "wss://")
            .Replace("http://", "ws://");
        wsUrl = $"{wsUrl}/process/pty/{Uri.EscapeDataString(options.Id)}/connect";

        var ws = new ClientWebSocket();
        ws.Options.SetRequestHeader("Authorization", $"Bearer {_sandbox.ApiKey}");
        await ws.ConnectAsync(new Uri(wsUrl), ct);

        return new PtyHandle(
            ws,
            options.Id,
            (cols, rows) => ResizePtySessionAsync(options.Id, cols, rows),
            () => KillPtySessionAsync(options.Id),
            options.OnData
        );
    }

    public Task ResizePtySessionAsync(string sessionId, int cols, int rows, CancellationToken ct = default)
        => GeneratedClientSupport.ExecuteToolboxAsync(
            () => _api.ResizePtySessionAsync(sessionId, new PtyResizeRequestModel(cols, rows), ct),
            ct
        );

    public Task KillPtySessionAsync(string sessionId, CancellationToken ct = default)
        => GeneratedClientSupport.ExecuteToolboxAsync(
            () => _api.DeletePtySessionAsync(sessionId, ct),
            ct
        );

    private static string MergeEnvironmentWithCommand(string command, Dictionary<string, string>? env)
    {
        if (env is null || env.Count == 0)
        {
            return command;
        }

        var envPrefix = string.Join(' ', env.Select(kv => $"{kv.Key}={EscapeShellArg(kv.Value)}"));
        return $"{envPrefix} {command}";
    }

    private static string EscapeShellArg(string value)
    {
        var escaped = value.Replace("'", "'\"'\"'");
        return $"'{escaped}'";
    }

    private static (string Stdout, string Stderr) DemuxOutput(string output)
    {
        var bytes = Encoding.UTF8.GetBytes(output);
        var stdout = new MemoryStream();
        var stderr = new MemoryStream();
        var target = stdout;

        var i = 0;
        while (i < bytes.Length)
        {
            if (HasPrefix(bytes, i, StdoutPrefix))
            {
                target = stdout;
                i += StdoutPrefix.Length;
                continue;
            }

            if (HasPrefix(bytes, i, StderrPrefix))
            {
                target = stderr;
                i += StderrPrefix.Length;
                continue;
            }

            target.WriteByte(bytes[i]);
            i++;
        }

        return (Encoding.UTF8.GetString(stdout.ToArray()), Encoding.UTF8.GetString(stderr.ToArray()));
    }

    private static bool HasPrefix(byte[] source, int offset, byte[] prefix)
    {
        if (offset + prefix.Length > source.Length)
        {
            return false;
        }

        for (var i = 0; i < prefix.Length; i++)
        {
            if (source[offset + i] != prefix[i])
            {
                return false;
            }
        }

        return true;
    }
}